src/Foundation/NSUrlSessionHandler.cs

```diff
@@ -538,7 +539,7 @@ public override int Read (byte [] buffer, int offset, int count)
 				if (d.Position == d.Length) {
 					lock (dataLock) {
 						// this is the same object, it was done to make the cleanup
-						data.Dequeue ();
+						data.Dequeue ().Dispose ();
```

Comment:
The reverse of `DangerousRetain` is not `Dispose`, it's `DangerousRelease`.

This change effectively leaks all native NSData objects.

In any case this change looks wrong, the managed NSData instance should retain the native object, which means this fix is just hiding the real other problem.

Reply:
👍 on `DangerousRelease`

Reply:
What I recall from when I looked at the bug (xmas break) was that `NSData` instance were _reused_ (by design or by a bug) and the dequeue was breaking this (since we killed the managed instance). 

It's possible the extra retain either block the behaviour (if by design) or workaround the bug.

Reply:
@spouliot it is working around the bug, but I'm trying to find a cleaner solution over this Retain approach. It is a workaround/fix until I find a better one.